### PR TITLE
testing(CPA): Unit tests added for CPA lib data-access services. 

### PR DESCRIPTION
**Pull request by IsaiahV98**
The following PR was for getting back into the groove of testing. Establishing how to test services with dependencies on other services and Http Client. 

### **Process for testing a service dependent on another service:** 
1. Create a skeleton mock. 
`const spyForDependentService= { someFunction: jest.fn() }`
 2. In the beforeEach establish the providers array:
```
beforeEach(() => {
  TestBed.configureTestingModule({
    providers: [
      ServiceToBeTested,
        {
          provide: DependentService,
          useValue: spyForDependentService
        }
      ]
    });
});
```
3. If the logic of the service has different actions based on a value condition of the dependent service, don't inject the service being tested in the beforeEach configureTestingModule statement. Instead inject it in the individual test cases after mocking the return value for said dependent services function call. 

### **Process for testing a HttpClient request (based on https://angular-training-guide.rangle.io/testing/services/http/using-httptestingmodule):** 
1. Import HttpClientTestingModule and HttpTestingController
` import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';`
2. Declare HttpTestingController in the outer describe statement, import HttpClientTestingModule in the configureTestingModule statement, and inject the HttpTestingController into the testbed. 
```
beforeEach(() => {
  TestBed.configureTestingModule({
    imports: [HttpClientTestingModule],
    providers: [
      ServiceToBeTested,
    ]
  });
  httpTestingController = TestBed.inject(HttpTestingController);
});
```
3. Make a dummy response and value, call the function in the service to be tested that makes a HttpClient call, subscribe to it, listen for specific response. 
```
const dummyValue = whatever value needed for function call;
const dummyResponse = [
  {
    whateverneeded: 'Foo',
  }
];

service.functionbeingtested(dummyValue).subscribe((interface/object) => {
  expect(interface/object.whateverneeded).toBe('Foo');
});
```
4. Call the httpTestingController with the expected API url the response will be made on, test the http method, and flush the response. 
```
const req = httpTestingController.expectOne(APIURL);
expect(req.request.method).toEqual('GET/POST/Whatever');
req.flush(dummyResponse);
```

### CODE REVIEW

File: libs/cpa/ngx/data-access/src/lib/services/response/response.service.spec.ts [hunk 1]
```diff
@@ -1,25 +1,53 @@
...
+  });
+
+  it('createResponse()', () => {
+    const dummyStringValue = 'Foo';
+    const dummyResponse = [
+      {
+        name: 'Foo',
+        workshops: []
+      }
+    ];
```
Comment (Proantagonist) [inline libs/cpa/ngx/data-access/src/lib/services/response/response.service.spec.ts:40]: Incorrect input and return type. I realize the service method being tested doesn't have a very useful implicit return type, but you are encouraged to look at the implementation and determine what kind of response comes back, if any, and have more accurate tests.

File: libs/cpa/ngx/data-access/src/lib/services/scenario/scenario.service.spec.ts [hunk 2]
```diff
@@ -1,25 +1,52 @@
...
+        workshops: []
+      }
+    ];
+
+    spyEnvironment.value.mockReturnValue(dummyStringValue);
+    service = TestBed.inject(ScenarioService);
+
+    service.create({}).subscribe((response) => {
+      expect(response[0].name).toBe('Foo');
+    });
```
Comment (Proantagonist) [inline libs/cpa/ngx/data-access/src/lib/services/scenario/scenario.service.spec.ts:46]: Incorrect return type and incorrect assertion

File: libs/cpa/ngx/data-access/src/lib/services/scenario/scenario.service.spec.ts [hunk 3]
```diff
@@ -1,25 +1,52 @@
...
 
-  it('should be created', () => {
-    const service: ScenarioService = TestBed.get(ScenarioService);
-    expect(service).toBeTruthy();
+    const req = httpTestingController.expectOne(`${'Fooscenarios'}`);
```
Comment (Proantagonist) [inline libs/cpa/ngx/data-access/src/lib/services/scenario/scenario.service.spec.ts:48]: This is definitely not the endpoint

File: libs/cpa/ngx/data-access/src/lib/services/snapshot/snapshot.service.spec.ts [hunk 4]
```diff
@@ -1,25 +1,53 @@
...
+        workshops: []
+      }
+    ];
+
+    spyEnvironment.value.mockReturnValue(dummyStringValue);
+    service = TestBed.inject(SnapshotService);
+
+    service.create({}).subscribe((response) => {
+      expect(response[0].name).toBe('Foo');
+    });
```
Comment (Proantagonist) [inline libs/cpa/ngx/data-access/src/lib/services/snapshot/snapshot.service.spec.ts:47]: Invalid return type and assertion

File: libs/cpa/ngx/data-access/src/lib/services/workshop/workshop.service.spec.ts [hunk 5]
```diff
@@ -1,25 +1,53 @@
...
+        workshops: []
+      }
+    ];
+
+    spyEnvironment.value.mockReturnValue(dummyStringValue);
+    service = TestBed.inject(WorkshopService);
+
+    service.createWorkshop({ guid: 'bruh' }).subscribe((response) => {
+      expect(response[0].name).toBe('Foo');
+    });
```
Comment (Proantagonist) [inline libs/cpa/ngx/data-access/src/lib/services/workshop/workshop.service.spec.ts:47]: Incorrect return type and assertion

**Comment by Proantagonist** [comment]
Discussion:

If we can write tests with any random input values, endpoint urls, etc, and mock any kind of response, and make assertions that effectively are `true == true`, then what is the value of these tests. I have a really hard time understanding why we should dedicate any amount of time to these tests except for maybe coverage reporting.

**Review by Proantagonist — CHANGES_REQUESTED** [review]

File: libs/cpa/ngx/data-access/src/lib/services/participant/participants.service.spec.ts [hunk 6]
```diff
@@ -1,16 +1,75 @@
 import { TestBed } from '@angular/core/testing';
+import { HttpClientTestingModule, HttpTestingController } from '@angular/common/http/testing';
 
+import { EnvironmentService } from '@tamu-gisc/common/ngx/environment';
 import { ParticipantService } from './participants.service';
 
+const spyEnvironment = {
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/cpa/ngx/data-access/src/lib/services/participant/participants.service.spec.ts:7]: Jest as a testing library makes no real distinction between spys and mocks (both result in mock functions): https://jestjs.io/docs/jest-object#jestspyonobject-methodname

**Comment by IsaiahV98** [comment]
> Discussion:
> 
> If we can write tests with any random input values, endpoint urls, etc, and mock any kind of response, and make assertions that effectively are `true == true`, then what is the value of these tests. I have a really hard time understanding why we should dedicate any amount of time to these tests except for maybe coverage reporting.

The value in these tests came from me attempting to figure how to deal with angular service dependencies utilizing Jest. Identifying and documenting the correct way deal with scenarios in which a service is dependent on another service, and identifying and documenting the correct way to deal with scenarios in which a service is dependent on Angular’s  HttpClient via the utilization of the HttpClientTestingModule and HttpTestingController. The files and corresponding functions under test here have their logic completely dependent on dependencies, which means to my understanding that there is basically nothing to test for in the scope of a unit test. So, for future testing there is no point outside of increasing testing coverage for unit testing these kinds of files/functions. 

However, now that it is documented how to properly utilize HttpClientTestingModule and HttpTestingController, we can in the future refer back to this documentation and code example for scenarios in which logic is not completely dependent on dependencies. Such as pipe calls on HttpClient requests and such.

File: libs/cpa/ngx/data-access/src/lib/services/participant/participants.service.spec.ts [hunk 7]
```diff
@@ -1,16 +1,75 @@
...
+  });
+
+  it('createParticipantForWorkshop()', () => {
+    const dummyStringValue = 'Foo';
+    const dummyResponse = [
+      {
+        name: 'Foo',
+        workshops: []
+      }
+    ];
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/cpa/ngx/data-access/src/lib/services/participant/participants.service.spec.ts:40]: > This is also where my understanding of the purpose of these tests falls apart. The dummyResponse is the incorrect return type for the service call createParticipantForWorkshop. The service method returns an object representing a participant, not an array of participants which is what you're testing for.

This is something I should have payed closer attention to, and will work on fixing.

File: libs/cpa/ngx/data-access/src/lib/services/participant/participants.service.spec.ts [hunk 8]
```diff
@@ -1,16 +1,75 @@
...
+      ]
+    });
+    httpTestingController = TestBed.inject(HttpTestingController);
+  });
+
+  afterEach(() => {
+    httpTestingController.verify();
+  });
+
+  it('createParticipantForWorkshop()', () => {
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/cpa/ngx/data-access/src/lib/services/participant/participants.service.spec.ts]: Fixed

File: libs/cpa/ngx/data-access/src/lib/services/participant/participants.service.spec.ts [hunk 9]
```diff
@@ -1,16 +1,75 @@
...
+
+    spyEnvironment.value.mockReturnValue(dummyStringValue);
     service = TestBed.inject(ParticipantService);
+
+    service.createParticipantForWorkshop(dummyStringValue, dummyStringValue).subscribe((participants) => {
+      expect(participants[0].name).toBe('Foo');
+    });
+
+    const req = httpTestingController.expectOne(`${service.resource}/workshop`);
+    expect(req.request.method).toEqual('POST');
```
[the remark this answers is not in the document]
Comment (IsaiahV98) [reply libs/cpa/ngx/data-access/src/lib/services/participant/participants.service.spec.ts]: Works, and applied for all tests.

**Pull request merged**